I3uckwheat/cards-of-carousal
# Feature/207 loading states

#### 0. Make sure your issue is linked:


Closes #228 

#### 1. Purpose:


The purpose of this submission is to implement various loading states into the application (ex. Displaying a loading indicator when joining a lobby). I haven't covered every possible case of loading states, but I've done the major ones:

- player connecting to host
- submitting cards to host
- getting card packs
- getting cards
- getting new join-code

#### 2. Implementation:


The major changes that were made in order to make this work were:

- Adding a `loading` value to the Player and Host contexts. `loading` is an array of strings, and the strings represent what is currently loading in the application
- Updating reducers to add/remove strings from the `loading` array
- Updating components to display loading indicators when a certain value is present in the `loading` array

For a more detailed look at loading states, I've written some documentation https://github.com/I3uckwheat/cards-of-carousal/wiki/Loading-States

#### 3. Possible Issues:


I've written about some special cases in my documentation, but it's possible I may have missed something.

#### 4. How to Test:


Checkout my branch and pull any changes

- player connecting to host
  - `frontend/src/socket/socket.js`: Comment out line 33 where the message is being sent, and un-comment the comment above it which contains the setTimeout
  - Now when you start a game and join as a player, you should see a loading indicator
- submitting cards to host
  - Follow the steps in [this PR](https://github.com/I3uckwheat/cards-of-carousal/pull/181) so you can submit cards
  - Make sure the setTimeout is un-commented in `socket.js` and submit your cards. You should see a loading indicator
- getting card packs
  - `GameSettings.js`: Wrap the `PACKS_RECEIVED` dispatch in a setTimeout with a value of around 1 second so you can see the loading indicator
  - Start a game as a host and you should see the loading indicator where the pack names should be
- getting cards
  - Start a game as a host, join the game as a player, then on the host screen select some card packs and hit the START CAROUSING button. You should see a quick glimpse of the loading indicator in the center of the right panel of the host screen (not sure how to setTimeout this one)
- getting new join-code
  - Add the setTimeout to the socket like we did earlier and hit Shuffle Join Code in the host settings. There should be a loading indicator in place of the join code while it loads


#### 5. Cleanup
